### ISSUE TITLE
Proposal: Don't mutate objects in common hooks

### ISSUE THREAD

**Issue by daffl** [feathersjs-ecosystem/feathers-hooks-common#480]
This is kind of a follow up to https://github.com/feathers-plus/feathers-hooks-common/pull/479:

Mutating objects, especially deleting from existing objects is probably one of the biggest challenge when dealing with hooks. For example:

- Removing or adding to `params` or `params.query` and then passing it to another hook will cause really confusing and hard to debug problems
- When calling methods internally, mutating e.g. `context.data` especially in a hook can cause even more problems, for example:

```js
const data = { text: 'hi' };

app.service('messages').create(data);

// data.userId is now set. This is not what I'd expect
```

I mostly changed the approach in my hooks to replace `query`, `params` and `data` with shallow or - if appropriate - full copies and was able to avoid many of those issues. The changes I am proposing for the common hooks would be:

- Use [lodash.omit](https://lodash.com/docs/4.17.11#omit) instead of `delete-by-dot` since it does not mutate the original object and replace the appropriate part on the context
- Create a `before` hook that makes a copy of `context.params`, `context.params.query` and `context.data`

**Comment by eddyystop** [comment]
Noted.

**Comment by daffl** [comment]
This has been mostly addressed by #543
